stetho/src/main/java/com/facebook/stetho/inspector/network/DownloadingAsyncPrettyPrinterFactory.java

```diff
@@ -0,0 +1,170 @@
+          return result.getDisplayType();
+        }
+      };
+    }
+  }
+
+  /**
+   * Match the correct header that contains information about the schema uri
+   * @param headerName header name of a response that needs to be pretty printed
+   * @param headerValue header value which contains the URI for
+   * the schema data needed to pretty print the response body
+   * @return MatchResult that has the schema uri and the type of prettified result. Null
+   * if there is no correct header match.
+   */
+  @Nullable
+  protected abstract MatchResult matchAndParseHeader(String headerName, String headerValue);
+
+  protected abstract void doPrint(PrintWriter output, InputStream payload, String schema)
+      throws IOException;
+
+  @Nullable
+  private static URL parseURL(String uri) {
+    try {
+      return new URL(uri);
+    } catch (MalformedURLException e) {
+      return null;
+    }
+  }
+
+  private static void doErrorPrint(PrintWriter output, InputStream payload, String errorMessage)
+    throws IOException {
+    output.print(errorMessage + "\n" + Util.readAsUTF8(payload));
+  }
+
+  private static AsyncPrettyPrinter getErrorAsyncPrettyPrinter(
+      final String headerName,
+      final String headerValue) {
+    return new AsyncPrettyPrinter() {
+      @Override
+      public void printTo(PrintWriter output, InputStream payload) throws IOException {
+        String errorMessage = "[Failed to parse header: "
+            + headerName + " : " + headerValue + " ]";
+        doErrorPrint(output, payload, errorMessage);
+      }
+
+      @Override
+      public PrettyPrinterDisplayType getPrettifiedType() {
+        return PrettyPrinterDisplayType.TEXT;
+      }
+    };
+  }
+
+
+  public enum PrettyPrinterDisplayType {
+    JSON("Json"),
+    HTML("Html"),
+    TEXT("Text");
+
+    private final String mDisplayType;
+
+    private PrettyPrinterDisplayType(String displayType) {
+      mDisplayType = displayType;
+    }
+
+    public String getDisplayType() {
+      return mDisplayType;
+    }
+  }
+
+  protected class MatchResult {
+    private String mSchemaUri;
+    private PrettyPrinterDisplayType mDisplayType;
+
+    public MatchResult(String schemaUri, PrettyPrinterDisplayType displayType) {
+      mSchemaUri = schemaUri;
+      mDisplayType = displayType;
+    }
+
+    public String getSchemaUri() {
+      return mSchemaUri;
+    }
+
+    public PrettyPrinterDisplayType getDisplayType() {
+      return mDisplayType;
+    }
+  }
+
+  private class Request implements Callable<String> {
+    private URL url;
+
+    public Request(URL url) {
+      this.url = url;
+    }
+
+    @Override @Nullable
+    public String call() {
+      try {
+        InputStream urlStream = url.openStream();
+        String result = Util.readAsUTF8(urlStream);
+        urlStream.close();
+        return result;
+      }
+      catch (IOException e) {
+        return null;
```

Comment:
Your code above pretends to handle this case, and yet you just return null instead, sacrificing all of the error information you previously had.  This is the kind of thing that makes engineers tear their hair out during multi-hour debug sessions trying to find this line.

Instead, declare that you throw an `IOException` and let it bubble up at least to the anonymous class in `getInstance`.  FWIW, I think the better design is to let it bubble all the way up to `Network#getResponseBody`  where it would actually yield an error to the chrome devtools, allowing it to potentially be retried if the user clicks again, but at this point that should be looked at as a follow-up to your base work.